src/main/java/org/apache/sysml/runtime/instructions/cp/ParamservBuiltinCPInstruction.java

```diff
@@ -114,33 +116,67 @@ public void processInstruction(ExecutionContext ec) {
+		Timing tSetup = DMLScript.STATISTICS ? new Timing(true) : null;
+
 		PSScheme scheme = getScheme();
 		int workerNum = getWorkerNum(mode);
 		String updFunc = getParam(PS_UPDATE_FUN);
 		String aggFunc = getParam(PS_AGGREGATION_FUN);
 
-		int k = getParLevel(workerNum);
-
 		// Get the compiled execution context
 		LocalVariableMap newVarsMap = createVarsMap(sec);
-		ExecutionContext newEC = ParamservUtils.createExecutionContext(sec, newVarsMap, updFunc, aggFunc, k);
+		ExecutionContext newEC = ParamservUtils.createExecutionContext(sec, newVarsMap, updFunc, aggFunc, 1);	// level of par is 1 in spark backend
 
 		MatrixObject features = sec.getMatrixObject(getParam(PS_FEATURES));
 		MatrixObject labels = sec.getMatrixObject(getParam(PS_LABELS));
 
 		// Force all the instructions to CP type
-		Recompiler.recompileProgramBlockHierarchy2Forced(
-			newEC.getProgram().getProgramBlocks(), 0, new HashSet<>(), LopProperties.ExecType.CP);
-		
+		Recompiler.recompileProgramBlockHierarchy2Forced(newEC.getProgram().getProgramBlocks(), 0, new HashSet<>(), LopProperties.ExecType.CP);
+
 		// Serialize all the needed params for remote workers
 		SparkPSBody body = new SparkPSBody(newEC);
 		HashMap<String, byte[]> clsMap = new HashMap<>();
 		String program = ProgramConverter.serializeSparkPSBody(body, clsMap);
 
-		SparkPSWorker worker = new SparkPSWorker(getParam(PS_UPDATE_FUN), getFrequency(), getEpochs(), getBatchSize(), program, clsMap);
-		ParamservUtils.doPartitionOnSpark(sec, features, labels, scheme, workerNum) // Do data partitioning
-			.foreach(worker);   // Run remote workers
+		// Get some configurations
+		String host = sec.getSparkContext().getConf().get("spark.driver.host");
+		long rpcTimeout;
+		if (sec.getSparkContext().getConf().contains("spark.rpc.askTimeout")) {
+			rpcTimeout = sec.getSparkContext().getConf().getTimeAsMs("spark.rpc.askTimeout");
+		} else {
+			rpcTimeout = sec.getSparkContext().getConf().getTimeAsMs("spark.network.timeout", "120s");
+		}
 
+		// Create remote workers
+		SparkPSWorker worker = new SparkPSWorker(getParam(PS_UPDATE_FUN), getParam(PS_AGGREGATION_FUN), getFrequency(),
+			getEpochs(), getBatchSize(), program, clsMap, host, rpcTimeout);
+
+		// Create the agg service's execution context
+		ExecutionContext aggServiceEC = ParamservUtils.copyExecutionContext(newEC, 1).get(0);
+
+		// Create the parameter server
+		ListObject model = sec.getListObject(getParam(PS_MODEL));
+		ParamServer ps = createPS(mode, aggFunc, getUpdateType(), workerNum, model, aggServiceEC);
+
+		if (DMLScript.STATISTICS)
+			Statistics.accPSSetupTime((long) tSetup.stop());
+
+		// Create the netty server for ps
+		TransportServer server = PSRpcFactory.createServer((LocalParamServer) ps, host);	// Start the server
+
+		try {
+			ParamservUtils.doPartitionOnSpark(sec, features, labels, scheme, workerNum) // Do data partitioning
+			    .foreach(worker);   // Run remote workers
```

Comment:
Please use accumulators to collect the statistics (that we show on -stats) from all executor processes unless we're running in local spark mode as indicated by `SparkExecutionContext.isLocalMaster()`  (parfor can serve as an example). Right now the statistics collection would only work in local mode where all "remote" tasks are executed in the driver process and hence correctly update the static statistics. In cluster mode, all statistics from the executors would be lost. 

Also, it might be a good idea to use additional accumulators for the number of executed batches and epochs. This gives users a rough progress indicator in the Spark UI which will be very useful for long running paramserv instances.